gammapy/spectrum/energy.py

```diff
@@ -0,0 +1,293 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+from __future__ import print_function, division
+from astropy.units import Quantity
+from astropy.io import fits
+from ..utils.array import array_stats_str
+
+import astropy.units as u
+import numpy as np
+
+__all__ = ['Energy',
+           'EnergyBounds']
+
+class Energy(u.Quantity):
+
+    """Energy bin centers.
+
+    Stored as "ENERGIES" FITS table extensions.
+
+    Parameters
+    ----------
+    energy : `~numpy.array`, scalar, `~astropy.units.Quantity`
+        Energy
+
+    unit : `~astropy.units.UnitBase`, str, optional
+        The unit of the value specified for the energy.  This may be
+        any string that `~astropy.units.Unit` understands, but it is
+        better to give an actual unit object.
+
+    dtype : `~numpy.dtype`, optional
+        See `~astropy.units.Quantity`.
+
+    copy : bool, optional
+        See `~astropy.units.Quantity`.
+
+    """
+
+    def __new__(cls, energy, unit=None, dtype=None, copy=True):
+
+        # Techniques to subclass Quantity taken from astropy.coordinates.Angle
+        # see: http://docs.scipy.org/doc/numpy/user/basics.subclassing.html
+        if unit is not None:
+            unit = u.Unit(unit)
+            if not unit.is_equivalent(u.eV):
+                raise ValueError("Requested unit {0} is not an"
+                                 " energy unit".format(unit))
+
+        if isinstance(energy, u.Quantity):
+            # also True if energy is of type Energy
+            if unit is not None:
+                energy = energy.to(unit).value
+            else:
+                unit = energy.unit
+                if not unit.is_equivalent(u.eV):
+                    raise ValueError("Given quantity {0} is not an"
+                                     " energy".format(energy))
+                energy = energy.value
+
+        else:
+            if unit is None:
+                raise ValueError("No unit given")
+
+        self = super(Energy, cls).__new__(cls, energy, unit,
+                                          dtype=dtype, copy=copy)
+
+        # Interesting once  energy bounds are stored
+        self._nbins = self.size
+
+        return self
+
+    def __array_finalize__(self, obj):
+        if obj is None:
+            return
+        self._nbins = self.size
```

Comment:
If you use `self.size` in the `nbins` property, you do not need to redefine `__array_finalize__` at all. If you keep it as is, I'd suggest to use call `super(Energy, self).__array_finalize__()`, just in case we at some point change what `Quantity` does (or if someone decides, e.g., to define a new `class MyEnergy(Energy, MyQuantity)`).